clearnode/app_session_service.go

```diff
@@ -134,66 +137,44 @@ func (s *AppSessionService) SubmitAppState(params *SubmitAppStateParams, rpcSign
 		sessionAccountID := NewAccountID(appSession.SessionID)
 
 		newVersion = appSession.Version + 1
-		switch appSession.Protocol {
-		case rpc.VersionNitroRPCv0_2:
-			// nothing additional to check
-		case rpc.VersionNitroRPCv0_4:
-			if newVersion != params.Version {
-				return RPCErrorf("invalid version: expected %d, got %d", newVersion, params.Version)
-			}
-
-			switch params.Intent {
-			case rpc.AppSessionIntentOperate:
-				// no additional actions needed
-			default:
-				return RPCErrorf("unsupported intent: %s", params.Intent)
-			}
-		default:
-			return RPCErrorf("unsupported protocol: %s", appSession.Protocol)
-		}
 
 		appSessionBalance, err := getAppSessionBalances(tx, sessionAccountID)
 		if err != nil {
 			return err
 		}
 
-		allocationSum := map[string]decimal.Decimal{}
-		for _, alloc := range params.Allocations {
-			if alloc.Amount.IsNegative() {
-				return RPCErrorf("negative allocation: %s for asset %s", alloc.Amount, alloc.AssetSymbol)
+		switch {
+		case appSession.Protocol == rpc.VersionNitroRPCv0_4:
+			if newVersion != params.Version {
+				return RPCErrorf("invalid version: expected %d, got %d", newVersion, params.Version)
 			}
-
-			walletAddress := GetWalletBySigner(alloc.ParticipantWallet)
-			if walletAddress == "" {
-				walletAddress = alloc.ParticipantWallet
+			switch params.Intent {
+			case rpc.AppSessionIntentDeposit:
+				depositParticipants, err := s.handleDepositIntent(tx, appSession, params, rpcSigners, sessionAccountID, appSessionBalance)
+				if err != nil {
+					return err
+				}
+				for participant := range depositParticipants {
+					participants[participant] = true
+				}
+			case rpc.AppSessionIntentOperate:
+				err := s.handleOperateIntent(tx, params, participantWeights, sessionAccountID, appSessionBalance)
+				if err != nil {
+					return err
+				}
+			default:
+				return RPCErrorf("unsupported intent: %s", params.Intent)
 			}
-
-			if _, ok := participantWeights[walletAddress]; !ok {
-				return RPCErrorf("allocation to non-participant %s", walletAddress)
+		case appSession.Protocol == rpc.VersionNitroRPCv0_2:
+			if params.Intent == rpc.AppSessionIntentDeposit {
```

Comment:
Here I suppose we can even reject if either intent or version is not empty, so users don't think it will work